File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [0]*len(nums)
        zero_cnt = 0
        zero_pos = 0
        product_all = 1

        # More than 2 zero : return all zero
        # 1 zero : Find zero position and put calculated value for zero position in result array
        # Non zero : divide product all by each num

        for id, val in enumerate(nums):
            if val == 0:
                zero_cnt += 1

                if zero_cnt > 1:
                    return ans
                zero_pos = id
            else:
                product_all *= val

        if zero_cnt==1:
            ans[zero_pos] = product_all
            return ans
        
        for index in range(len(nums)):
            ans[index] = product_all//nums[index]

        return ans

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int left[n];
        int right[n];

        left[0]=1;
        for(int i=1; i<nums.size(); i++) {
            left[i] = nums[i-1]*left[i-1];
        }
        right[nums.size()-1] = 1;
        for(int i=nums.size()-2; i>=0; i--) {
            right[i] = nums[i+1]*right[i+1];
        }

        vector<int> ret(nums.size());
        for(int i=0; i<nums.size(); i++) {
            ret[i] = left[i]*right[i];
        }

        return ret;
    }
};

/*
O(n) time without division
32 bit
[2,2,3,4]
[24,24,16,12]

  .... i ....
  all product of i's left, i , all product of i's right

left = [1,]
left[i] = left[i-1]*nums[i-1]
right = [1,]
right[i] = right[r+1]*nums[i+1]


left = [1,]
left[i] = left[i-1]*nums[i-1]

right = 1
for(i=size-1; i>=0 i--) {
    left[i] = left[i] * right
    right *= nums[i] // right will have all product of i's right
}
*/